src/storage.rs

```diff
@@ -2390,23 +2276,359 @@ impl Cache {
+
+        // Update the users cache and individual crates caches, and return our
+        // set of versions.
+        let mut guard = self.cache.state.lock().unwrap();
+        let versions: SortedMap<_, _> = result
+            .versions
+            .into_iter()
+            .map(|api_version| {
+                (
+                    api_version.num,
+                    Some(CratesCacheVersionDetails {
+                        created_at: api_version.created_at,
+                        published_by: api_version.published_by.map(|api_user| {
+                            info!("recording user info for {api_user:?}");
+                            guard.crates_cache.users.insert(
+                                api_user.id,
+                                CratesCacheUser {
+                                    login: api_user.login,
+                                    name: api_user.name,
+                                },
+                            );
+                            api_user.id
+                        }),
+                    }),
+                )
+            })
+            .collect();
+        info!(
+            "found {} versions for crate {}",
+            versions.len(),
+            self.crate_name
+        );
+        guard.crates_cache.crates.insert(
+            self.crate_name.to_owned(),
+            CratesCacheEntry {
+                last_fetched: self.now,
+                versions,
+                metadata: Some(result.crate_data),
+            },
+        );
+
+        Ok(guard)
+    }
+
+    /// Use `index.crates.io` to get crate information.
+    ///
+    /// This will only add versions which aren't already present in `crates_cache.crates` for
+    /// `name`.
+    ///
+    /// When this function returns `Ok`, the returned state is guaranteed to have an entry for
+    /// `name` in `crates_cache.crates`.
+    async fn update_using_index(
+        &self,
+        network: &Network,
+    ) -> Result<std::sync::MutexGuard<'a, CacheState>, CrateInfoError> {
+        // Crate names can only be a subset of ascii (valid rust identifier characters and `-`), so
+        // using `len()` and indexing will result in valid counts/characters.
+        let mut url = String::from("https://index.crates.io/");
+        let name = self.crate_name;
+        use std::fmt::Write;
+        match name.len() {
+            1 => write!(url, "1/{}", name),
+            2 => write!(url, "2/{}", name),
+            3 => write!(url, "3/{}/{name}", &name[0..1]),
+            _ => write!(url, "{}/{}/{name}", &name[0..2], &name[2..4]),
+        }
+        .expect("writing to a String should not fail");
+        let url = Url::parse(&url).expect("invalid crate name");
+
+        let response = self.try_download(network, url).await?;
+
+        #[derive(Deserialize, Debug, Clone)]
+        struct IndexEntry {
+            vers: semver::Version,
+        }
+
+        let result = load_json_lines::<IndexEntry>(&response[..])?;
```

Comment:
You can get the full crates.io index values by using `crates_index::Crate::from_slice(...)` (https://docs.rs/crates-index/latest/crates_index/struct.Crate.html#method.from_slice), fwiw. So long as we're keeping that dependency around, there's not much point rolling our own deserialization.